src/Microsoft.AspNetCore.Blazor.Server/Circuits/RenderBatchFormatterResolver.cs

```diff
@@ -29,15 +34,54 @@ public RenderBatch Deserialize(byte[] bytes, int offset, IFormatterResolver form
 
             public int Serialize(ref byte[] bytes, int offset, RenderBatch value, IFormatterResolver formatterResolver)
             {
-                using (var memoryStream = new MemoryStream())
+                // WARNING: Be careful never to execute .GetBuffer() or .ToArray() on this memory
+                // stream we get from recyclableMemoryStreamManager. If you do that, you'll silently
+                // eliminate all the benefit of RecyclableMemoryStream.
+                using (var memoryStream = recyclableMemoryStreamManager.GetStream($"{nameof(RenderBatchFormatter)}.{nameof(Serialize)}"))
                 using (var renderBatchWriter = new RenderBatchWriter(memoryStream, leaveOpen: false))
                 {
                     renderBatchWriter.Write(value);
-                    
-                    var bytesBuffer = memoryStream.GetBuffer();
-                    return MessagePackBinary.WriteBytes(ref bytes, offset, bytesBuffer, 0, (int)memoryStream.Length);
+
+                    var compressedStream = CompressStream(memoryStream);
+                    try
+                    {
+                        return CopyStreamToMessagePack(compressedStream, ref bytes, offset);
+                    }
+                    finally
+                    {
+                        compressedStream.Dispose();
+                    }
+                }
+            }
+
+            private static Stream CompressStream(Stream input)
+            {
+                // WARNING: Be careful never to execute .GetBuffer() or .ToArray() on this memory
+                // stream we get from recyclableMemoryStreamManager. If you do that, you'll silently
+                // eliminate all the benefit of RecyclableMemoryStream.
+                var output = recyclableMemoryStreamManager.GetStream($"{nameof(RenderBatchFormatter)}.{nameof(CompressStream)}");
+
+                using (var gzipStream = new GZipStream(output, CompressionLevel.Optimal, leaveOpen: true))
+                {
+                    input.Seek(0, SeekOrigin.Begin);
+                    input.CopyTo(gzipStream);
+                    gzipStream.Flush();
+                    return output;
                 }
-                    
+            }
+
+            private static int CopyStreamToMessagePack(Stream stream, ref byte[] bytes, int offset)
+            {
+                var buffer = new byte[32768];
```

Comment:
Could we use `ArrayPool<>` here? It's fairly easy to use

Reply:
Sounds good - will look into it.

Reply:
Updated logic will eliminate the need for this temporary buffer.